Content.Shared/Timing/UseDelaySystem.cs

```diff
@@ -7,53 +8,154 @@ public sealed class UseDelaySystem : EntitySystem
     [Dependency] private readonly IGameTiming _gameTiming = default!;
     [Dependency] private readonly MetaDataSystem _metadata = default!;
 
-    public void SetDelay(Entity<UseDelayComponent> ent, TimeSpan delay)
+    private const string DefaultId = "default";
+
+    public override void Initialize()
     {
-        if (ent.Comp.Delay == delay)
-            return;
+        base.Initialize();
+
+        SubscribeLocalEvent<UseDelayComponent, MapInitEvent>(OnMapInit);
+        SubscribeLocalEvent<UseDelayComponent, EntityUnpausedEvent>(OnUnpaused);
+    }
+
+    private void OnMapInit(Entity<UseDelayComponent> ent, ref MapInitEvent args)
+    {
+        // Set default delay length from the prototype
+        // This makes it easier for simple use cases that only need a single delay
+        TryRegisterDelay(ent, DefaultId, ent.Comp.Delay);
+    }
 
-        ent.Comp.Delay = delay;
+    private void OnUnpaused(Entity<UseDelayComponent> ent, ref EntityUnpausedEvent args)
+    {
+        // We have to do this manually, since it's not just a single field.
+        foreach (var entry in ent.Comp.Delays.Values)
+        {
+            entry.EndTime += args.PausedTime;
+        }
+    }
+
+    /// <summary>
+    /// Adds an additional delay to be tracked by this component.
+    /// </summary>
+    /// <param name="id">identifier used to refer to the delay in other methods</param>
+    /// <param name="length">duration of the delay</param>
+    /// <returns></returns>
+    public bool TryRegisterDelay(Entity<UseDelayComponent> ent, string id, TimeSpan length)
+    {
+        // Make sure there's not already a delay registered with this ID
+        if (ent.Comp.Delays.ContainsKey(id))
+            return false;
+
+        ent.Comp.Delays.Add(id, new UseDelayInfo(length));
         Dirty(ent);
+        return true;
     }
 
     /// <summary>
-    /// Returns true if the entity has a currently active UseDelay.
+    /// Sets the length of the delay with the specified ID.
     /// </summary>
-    public bool IsDelayed(Entity<UseDelayComponent> ent)
+    public bool SetLength(Entity<UseDelayComponent> ent, TimeSpan length, string id = DefaultId)
     {
-        return ent.Comp.DelayEndTime >= _gameTiming.CurTime;
+        if (!ent.Comp.Delays.TryGetValue(id, out var entry))
+            return false;
+
+        if (entry.Length == length)
+            return true;
+
+        entry.Length = length;
+        Dirty(ent);
+        return true;
     }
```

Comment:
this API of needing to register a delay and then having a second function to set it separately is kinda weird.

I think it would make a lot more sense to just retain the old API of SetDelay and just have the optional id argument added in. Then, if someone tries to set a delay and it isn't registered, you simply register it then. Otherwise, you just set the value. Having two code paths which essentially do the same thing comes across as very weird and is probably going to be quite prone to bugs if people use it and don't necessarily understand it.

Reply:
I agree that would be better. I had it that way at first, then changed it... for some reason that I can't remember. I'll try changing it back and see if any issues come up.

Reply:
Yep, I killed TryRegisterDelay and switched to just registering with SetLength. Works fine.